server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -688,11 +695,12 @@ private static void parseNonDynamicArray(
         final String lastFieldName,
         String arrayFieldName
     ) throws IOException {
+        boolean supportStoringArrayOffsets = mapper != null && mapper.supportStoringArrayOffsets(context);
         String fullPath = context.path().pathAsText(arrayFieldName);
 
         // Check if we need to record the array source. This only applies to synthetic source.
         boolean canRemoveSingleLeafElement = false;
-        if (context.canAddIgnoredField()) {
+        if (context.canAddIgnoredField() && (parsesArrayValue(mapper) == false && supportStoringArrayOffsets == false)) {
```

Comment:
I think i spend 10+ minutes trying to understand this line. Let's at least comment what's going on. I wonder if we can structure this better. Can we maybe split parsing logic based on `parsesArrayValue` into different methods?

Reply:
Actually it looks like if `parsesArrayValue(mapper) == true` we won't even get here?